[Qt] show conflicts of unconfirmed transactions in the UI

jonasschnelli opened the pull request:
Currently, the GUI shows only conflicts with transaction in a block.

This results in displaying replaced transaction (RBF) wrong (the replaced transaction is no longer in the mempool, but still in the wallet-transactionlist, conflicts are not detected).

This PR will mark unconfirmed transactions that are **not** in our mempool and have conflicts as conflicted.

jonasschnelli committed:
[Qt] show mempool conflicts in the UI

On RPC level, we show possible mempool conflicts already via detecting them over the CWallet::GetConflicts() function.
Currently, the GUI shows only conflicts with transaction in a block.

This results in displaying replaced transaction (RBF) wrong.

This PR will mark unconfirmed transactions that are _not_ in our mempool and have conflicts as conflicted.

jonasschnelli commented:
Another option would be to hide these transactions completely.

jonasschnelli commented:
Before this PR:

Sending 49.9BTC and RBF with a 40.0BTC transaction.
<img width="1056" alt="bildschirmfoto 2016-04-06 um 17 17 40" src="https://cloud.githubusercontent.com/assets/178464/14321935/812c0b34-fc1b-11e5-900d-f1b0b2851347.png">

After this PR:
<img width="1056" alt="bildschirmfoto 2016-04-06 um 17 21 22" src="https://cloud.githubusercontent.com/assets/178464/14322116/01e1d920-fc1c-11e5-8914-759536647343.png">

MarcoFalke commented on src/qt/transactionrecord.cpp:
```diff
@@ -257,7 +261,21 @@ void TransactionRecord::updateStatus(const CWalletTx &wtx)
+            return true;
+    }
+
+    if (status.cur_num_blocks != chainActive.Height())
```
Nit: I think you can leave those three lines the way it was before: `return status.cur_num_blocks != chainActive.Height();`

jonasschnelli replied:
Right! Fixed.

MarcoFalke commented:
>  Another option would be to hide these transactions completely.

I don't think you want to hide wtx without telling the user/ without having the user told you to do so.

jonasschnelli commented:
> I don't think you want to hide wtx without telling the user/ without having the user told you to do so.

Yes. I agree. IMO the "mark as conflicted" solution is better.

MarcoFalke commented:
Hmm, so I have a non-rbf transaction back from august:

```
Status: 0/unconfirmed, not in memory pool
Date: 8/19/15 15:51
```

and sendrawt returns `Missing inputs (code -25)`, but it is not showing the new icon. Need to look into this...

jonasschnelli commented:
@MarcoFalke: does you non-rbf transaction back from august is in conflict with another transaction in your wallet? Only if it's so, the conflicted state icon will be shown.

MarcoFalke commented:
No, it didn't conflict within the wallet.

Tested ACK 614073f

MarcoFalke commented:
One could argue to use another status/icon for this e.g. https://raw.githubusercontent.com/jonasschnelli/bitcoin/47196712ec7f3ff2f5f814650d40366cb269fe56/src/qt/res/icons/transaction_replaceable.png with the conflicted icon (cross) in the lower left.

jonasschnelli commented:
Rebased and changed the icon for transactions with conflict but not being in the mempool (RBF)l:
<img width="962" alt="bildschirmfoto 2016-05-20 um 17 48 22" src="https://cloud.githubusercontent.com/assets/178464/15433629/65292948-1eb3-11e6-8846-22d28a6ed337.png">

luke-jr commented:
These seem to be two list items referring to the same transaction.

jonasschnelli commented:
@luke-jr: this PR does not focus on hiding out transactions from the wallet (a lot more complexity).
At the moment, a replaced RBF transaction is not listed as conflicted (https://github.com/bitcoin/bitcoin/pull/7826#issuecomment-206424972).

This PR is a tiny fix to improve the GUI RBF attribution.

jonasschnelli commented:
Another solution would be (idea once shortly discusses with @gmaxwell on IRC):
* directly hide transactions if they are conflicted and **not** in the mempool (RBF)
* hide transactions that conflicts with a transaction in a block >6
* add option in the settings panel to "display old conflicted transaction" and turn it off by default.

sipa commented:
@jonasschnelli Any progress here?

jonasschnelli commented:
@sipa: I'm working on a option to auto-hide mempool conflicts (transactions that are not in the blockchain and mempool) (RBF/abandoned txes).

jonasschnelli commented:
I think this solution is fine as it is. The auto-hide option can be written in a separate PR.
This PR would be a simple change to allow users to easy identify mempool replacments relevant to the wallet.

Would be great for 0.13.

laanwj commented:
Removing this from the 0.13 milestone as it missed the feature freeze.

> Would be great for 0.13.

Agree, would have been nice, just like #7817. I think we have a lack of GUI testers/reviewers, as useful GUI changes like this linger way too long.

jonasschnelli commented:
This (simple) PR looks for reviewers...

MarcoFalke commented:
I think the problem is that it may not be obvious for everyone  to set up  the appropriate transactions to test this. I think it may help to add "proposed steps to reproduce" for harder to test pulls in general.

MarcoFalke commented:
Works for standard transactions:

![screenshot from 2016-08-10 10-37-01](https://cloud.githubusercontent.com/assets/6399679/17547541/1811fd7c-5ee8-11e6-94f9-c71c969244be.png)
![screenshot from 2016-08-10 10-37-06](https://cloud.githubusercontent.com/assets/6399679/17547540/16dc6a32-5ee8-11e6-9eb5-261cf4691693.png)
![screenshot from 2016-08-10 10-37-14](https://cloud.githubusercontent.com/assets/6399679/17547539/154c647e-5ee8-11e6-8ca7-a35d94bee9b9.png)

Does not work for non-standard transactions:

![screenshot from 2016-08-10 10-34-42](https://cloud.githubusercontent.com/assets/6399679/17547551/2124f63a-5ee8-11e6-8571-ae2435e44dd4.png)
![screenshot from 2016-08-10 10-34-45](https://cloud.githubusercontent.com/assets/6399679/17547552/22a62bdc-5ee8-11e6-91f4-a3c171aa9118.png)

jonasschnelli commented:
Closing for now due to inactivity...
